src/Pact/Native.hs

```diff
@@ -514,24 +514,30 @@ filter' i as = argsError' i as
 length' :: RNativeFun e
 length' _ [TList ls _ _] = return $ toTerm (length ls)
 length' _ [TLitString s] = return $ toTerm (T.length s)
-length' _ [TObject ps _ _] = return $ toTerm (length ps)
+length' _ [TObject (Object ps _ _) _] = return $ toTerm (length ps)
 length' i as = argsError i as
 
 take' :: RNativeFun e
 take' _ [TLitInteger c',TList l t _] = return $ TList (tord take c' l) t def
 take' _ [TLitInteger c',TLitString l] = return $ toTerm $ pack $ tord take c' (unpack l)
-take' _ [l@TList {},TObject {..}] =
-  return $ toTObject _tObjectType def $ (`filter` _tObject) $ \(k,_) -> searchTermList k (_tList l)
+take' _ [TList {..},TObject Object {..} _] = asKeyList _tList >>= \l ->
+  return $ toTObject _oObjectType def $ (`filter` _oObject) $ \(k,_) -> elem k l
 
 take' i as = argsError i as
 
 drop' :: RNativeFun e
 drop' _ [TLitInteger c',TList l t _] = return $ TList (tord drop c' l) t def
 drop' _ [TLitInteger c',TLitString l] = return $ toTerm $ pack $ tord drop c' (unpack l)
-drop' _ [l@TList {},TObject {..}] =
-  return $ toTObject _tObjectType def $ (`filter` _tObject) $ \(k,_) -> not $ searchTermList k (_tList l)
+drop' _ [TList {..},TObject Object {..} _] = asKeyList _tList >>= \l ->
+  return $ toTObject _oObjectType def $ (`filter` _oObject) $ \(k,_) -> not $ elem k l
 drop' i as = argsError i as
 
+asKeyList :: [Term Name] -> Eval e [FieldKey]
+asKeyList = mapM $ \t -> case t of
+  TLitString k -> return $ FieldKey k
+  _ -> evalError (_tInfo t) "String required"
```

Comment:
We should probably be more specific: "Literal string required"

Reply:
Except in this case it's not, the list has been reduced.

Reply:
got it